internal/controllers/database/sync.go

```diff
@@ -66,26 +53,44 @@ func (r *Reconciler) Sync(ctx context.Context, ydbCr *v1alpha1.Database) (ctrl.R
 	if stop {
 		return result, err
 	}
+	stop, result, err = r.letStorageKnowAboutThisDatabase(ctx, &database)
+	if stop {
+		return result, err
+	}
 	auth, result, err := r.getYDBCredentials(ctx, &database)
 	if auth == nil {
 		return result, err
 	}
 
-	if !meta.IsStatusConditionTrue(database.Status.Conditions, TenantInitializedCondition) {
-		stop, result, err = r.setInitialStatus(ctx, &database)
-		if stop {
-			return result, err
+	initialized := meta.IsStatusConditionTrue(database.Status.Conditions, DatabaseTenantInitializedCondition)
+	currentlyPaused := meta.IsStatusConditionTrue(database.Status.Conditions, DatabasePausedCondition)
+
+	if !currentlyPaused { //nolint:nestif
+		if !initialized {
+			stop, result, err = r.setInitialStatus(ctx, &database)
+			if stop {
+				return result, err
+			}
 		}
 		stop, result, err = r.waitForStatefulSetToScale(ctx, &database)
 		if stop {
 			return result, err
 		}
-		stop, result, err = r.handleTenantCreation(ctx, &database, auth)
-		if stop {
-			return result, err
+
+		if !initialized {
+			stop, result, err = r.handleTenantCreation(ctx, &database, auth)
+			if stop {
+				return result, err
+			}
 		}
 	}
-	return ctrl.Result{Requeue: false}, nil
+
+	stop, result, err = r.handlePauseResume(ctx, &database)
```

Comment:
We could just move this code up a bit if change count of replicas to 0 instead of deletion StatefulSet? Reconcile logic should handle pause before another operation and just skip steps